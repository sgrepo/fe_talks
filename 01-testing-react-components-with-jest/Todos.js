import React from 'react'

class TodosContainer extends React.Component {
  constructor(props) {
    super(props)

    this.state = {
      todos: this.props.todos || [],
      input: ''
    }

    this.onChange = this.onChange.bind(this)
    this.onAddTodo = this.onAddTodo.bind(this)
    this.onMarkAsDone = this.onMarkAsDone.bind(this)
  }

  onChange (e) {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  onAddTodo () {
    const { input } = this.state

    this.setState({
      input: '',
      todos: [
        ...this.state.todos,
        {
          done: false,
          text: input
        },
      ],
    })
  }

  onMarkAsDone (index) {
    const { todos } = this.state

    this.setState({
      todos: [
        ...todos.map((item, i) => {
          if (i === index) {
            item.done = !item.done
          }

          return item
        }),
      ]
    })
  }

  render() {
    const { onAddTodo, onMarkAsDone, onChange } = this
    const { todos, input } = this.state

    return (
      <div id="todos-container">
        <div id="todos-input">
          <input
            type="text"
            name="input"
            value={input}
            onChange={onChange}
          />
          <button
            onClick={onAddTodo}
          >Add todo!</button>
        </div>
        <div id="todos-list">
          <ul>
            { todos.map((todo, i) => (
              <li
                onClick={() => onMarkAsDone(i)}
                style={todo.done ? { textDecoration: 'line-through' } : {}}
                key={i}>
                {todo.text}</li>
            )) }
          </ul>
        </div>
      </div>
    )
  }
}

export default TodosContainer
