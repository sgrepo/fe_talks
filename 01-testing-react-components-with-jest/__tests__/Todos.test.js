import React from 'react'
import renderer from 'react-test-renderer'

import Todos from '../Todos'

it('should render an empty component', () => {
  const tree = renderer.create(
    <Todos />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

it('should render a list of todos', () => {
  const todos = [
    {
      text: 'some todo',
      done: false
    },
    {
      text: 'other todo',
      done: true
    }
  ]

  const tree = renderer.create(
    <Todos todos={todos} />
  ).toJSON()

  expect(tree).toMatchSnapshot()
})

it('should add a todo to a list of todos', () => {
  const component = renderer.create(
    <Todos />
  )

  // First snapshot
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()

  // Select components
  const todoContainer = tree.children.filter(child => child.type === 'div')[0]
  const todoInput = todoContainer.children.filter(child => child.type === 'input')[0]
  const todoButton = todoContainer.children.filter(child => child.type === 'button')[0]

  // Add todo to list
  todoInput.props.onChange({ target: { name: 'input', value: 'some content' } })
  todoButton.props.onClick()

  // Create second snapshot
  tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
