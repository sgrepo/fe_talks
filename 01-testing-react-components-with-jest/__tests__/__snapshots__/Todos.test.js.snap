exports[`test should add a todo to a list of todos 1`] = `
<div
  id="todos-container">
  <div
    id="todos-input">
    <input
      name="input"
      onChange={[Function]}
      type="text"
      value="" />
    <button
      onClick={[Function]}>
      Add todo!
    </button>
  </div>
  <div
    id="todos-list">
    <ul />
  </div>
</div>
`;

exports[`test should add a todo to a list of todos 2`] = `
<div
  id="todos-container">
  <div
    id="todos-input">
    <input
      name="input"
      onChange={[Function]}
      type="text"
      value="" />
    <button
      onClick={[Function]}>
      Add todo!
    </button>
  </div>
  <div
    id="todos-list">
    <ul>
      <li
        onClick={[Function]}
        style={Object {}}>
        some content
      </li>
    </ul>
  </div>
</div>
`;

exports[`test should render a list of todos 1`] = `
<div
  id="todos-container">
  <div
    id="todos-input">
    <input
      name="input"
      onChange={[Function]}
      type="text"
      value="" />
    <button
      onClick={[Function]}>
      Add todo!
    </button>
  </div>
  <div
    id="todos-list">
    <ul>
      <li
        onClick={[Function]}
        style={Object {}}>
        some todo
      </li>
      <li
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "line-through",
          }
        }>
        other todo
      </li>
    </ul>
  </div>
</div>
`;

exports[`test should render an empty component 1`] = `
<div
  id="todos-container">
  <div
    id="todos-input">
    <input
      name="input"
      onChange={[Function]}
      type="text"
      value="" />
    <button
      onClick={[Function]}>
      Add todo!
    </button>
  </div>
  <div
    id="todos-list">
    <ul />
  </div>
</div>
`;
